### PR TITLE
MINOR: [Release] Update C++ and PyArrow versions for 15.0.0

### PR DESCRIPTION
**Pull request by amoeba**
Can't easily test this locally so sending a PR to see what CI looks like. Java changes are in a separate PR: https://github.com/apache/arrow-cookbook/pull/339.

@raulcd

### CODE REVIEW

File: python/requirements.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 Sphinx>=4.0.2
-pyarrow==14.0.0
+pyarrow==15.0.0
 pandas>=1.2.5
 opentelemetry-api>=1.0.0
-opentelemetry-sdk>=1.0.0
\ No newline at end of file
+opentelemetry-sdk>=1.0.0opentelemetry-sdk>=1.0.0
```
Comment (lidavidm) [inline python/requirements.txt:5]: Something got messed up here? Do we even use this file if CI still passes?

Comment (raulcd) [reply python/requirements.txt:5]: We do and pip seems to do some magic knowing the name:
```
 Collecting opentelemetry-sdk>=1.0.0opentelemetry-sdk>=1.0.0
  Downloading opentelemetry_sdk-1.22.0-py3-none-any.whl (105 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 105.6/105.6 KB 11.3 MB/s eta 0:00:00
```

**Comment by raulcd** [comment]
Sorry, I forgot to run the release update for the cookbooks for the 14.0.x patch releases and 15.0.0, did you use the `01-bump-versions.sh 14.0.0 15.0.0` script?
https://github.com/apache/arrow-cookbook#usage

File: python/requirements.txt (hunk 1 above)
Comment (amoeba) [reply python/requirements.txt:5]: I have no idea what happened and didn't see this when I sent in the PR. I'll push a fix.

Comment (amoeba) [reply python/requirements.txt]: Fixed. Thanks for catching the mistake.

**Comment by amoeba** [comment]
I didn't originally but did just now (`sh dev/release/01-bump-versions.sh 14.0.0 15.0.0`) and squashed everything into a single commit. I'll do that in the future.

**Comment by amoeba** [comment]
@davisusanibar can you look at the Java CI failures?

**Comment by davisusanibar** [comment]
> > @davisusanibar can you look at the Java CI failures?
> 
> Let me take a look

A Java nightly is trying to download artifacts from https://nightlies.apache.org/arrow/java, but org.apache.arrow:arrow-memory-netty-buffer-patch:jar:16.0.0-SNAPSHOT is not being uploaded.

This issue has been created to track this: https://github.com/apache/arrow-cookbook/issues/342. Let me check the activities needed to upload that to the nightly repository.

**Comment by davisusanibar** [comment]
CC @vibhatha @danepitkin

**Comment by davisusanibar** [comment]
PR is ready for review: https://github.com/apache/arrow/pull/39901

**Comment by amoeba** [comment]
@raulcd I think now that https://github.com/apache/arrow/pull/39901 the Java checks should pass, can you re-run them to check?

**Comment by raulcd** [comment]
There seems to still be some errors:
```
 SLF4J(W): See https://www.slf4j.org/codes.html#ignoredBindings for an explanation.
Exception java.lang.UnsatisfiedLinkError: /tmp/jnilib-9784155468287532443.tmp: /tmp/jnilib-9784155468287532443.tmp: undefined symbol: _ZTIN6google8protobuf7MessageE
```
I can't spend much time diagnosing at the moment, @davisusanibar any idea what is happening on the Java failure?

**Comment by davisusanibar** [comment]
> There seems to still be some errors:
> 
> ```
>  SLF4J(W): See https://www.slf4j.org/codes.html#ignoredBindings for an explanation.
> Exception java.lang.UnsatisfiedLinkError: /tmp/jnilib-9784155468287532443.tmp: /tmp/jnilib-9784155468287532443.tmp: undefined symbol: _ZTIN6google8protobuf7MessageE
> ```
> 
> I can't spend much time diagnosing at the moment, @davisusanibar any idea what is happening on the Java failure?

I am reviewing that . More details at: https://github.com/apache/arrow-cookbook/pull/339#issuecomment-1932429139

**Review by vibhatha — COMMENTED** [review]
@davisusanibar can we check whether the Java-Jar related scenario is resolved?

**Comment by vibhatha** [comment]
We should probably re-run the CIs after merging https://github.com/apache/arrow/pull/40015

**Comment by davisusanibar** [comment]
> We should probably re-run the CIs after merging [apache/arrow#40015](https://github.com/apache/arrow/pull/40015)

This PR depend of https://github.com/apache/arrow-cookbook/pull/339

**Comment by davisusanibar** [comment]
> > We should probably re-run the CIs after merging [apache/arrow#40015](https://github.com/apache/arrow/pull/40015)
> 
> This PR depend of #339

@amoeba It's ready for a rebase

**Comment by amoeba** [comment]
Thanks @davisusanibar. I rebased and will keep an eye on CI.

**Comment by amoeba** [comment]
Hi @raulcd, I think this is ready for a merge. The CI passes here and the Java CI is passing as of https://github.com/apache/arrow-cookbook/actions/runs/7865593306. Can you have a look?

**Review by raulcd — APPROVED** [review]
Thanks for following this up!

**Pull request merged**